File: c++/pro-con/ch3/3.3-various-data-structures/SimpleProblemWithIntegers-ByBIT.cpp
#include <algorithm>
using namespace std;
typedef long long ll;
const int DAT_SIZE = (1<<18) - 1;

const int MAX_N = 100000;
const int MAX_Q = 100000;

//INPUT
int N, Q;
int A[MAX_N];
char T[MAX_Q];
int L[MAX_Q], R[MAX_Q], X[MAX_Q];

// BIT
ll bit0[MAX_N+1], bit1[MAX_N+1];

// calc sum
ll sum(ll *b, int i){
  ll s = 0;
  while(i>0){
    s += b[i];
    i -= i & -i;
  }
  return s;
}

void add(ll *b, int i, int v){
  while(i <= N){
    b[i] += v;
    i += i & -i;
  }
}


void solve(){
  for(int i=1; i<=N; i++){
    add(bit0, i, A[i]);
  }
  for(int i=0; i<Q; i++){
    if(T[i] == 'C'){
      add(bit0, L[i], -X[i] * (L[i] - 1));
      add(bit1, L[i], X[i]);
      add(bit0, R[i]+1, X[i] * R[i]);
      add(bit1, R[i]+1, -X[i]);
    } else {
      ll res = 0;
      res += sum(bit0, R[i]) + sum(bit1, R[i]) * R[i];
      res -= sum(bit0, L[i]-1) + sum(bit1, L[i]-1) * (L[i]-1);
      printf("%lld\n", res);
    }
  }
}

File: python/pro-con-python/ch3/sec3_SegmentTree_RMQ_BIT/SimpleProblemWithIntegersByBIT.py
'''
Created on 2016/12/18

@author: kazuyoshi
'''
MAX_N=100

#BIT
bit0=[0 for i in range(MAX_N+1)]
bit1=[0 for i in range(MAX_N+1)]
N=0

def sum(bit, i): # sum of a1, a2, ..., a_i
    s=0
    while i>0:
        s += bit[i]
        i -= i& -i  ## or i = i&(i-1)
    return s

def add(bit, i, x,N): # add x to a_i
    while i<=N:
        bit[i] += x
        i += i& -i

def solve(A,N,Q,T,L,R,X):
    for i,a in enumerate(A[1:]):
        add(bit0,i+1,a,N)
        
    for (t,l,r,x) in zip(T,L,R,X):
        if t == "C":
            add(bit0, l, -x*(l-1),N)
            add(bit1, l, x,N)
            add(bit0, r+1, x*r,N)
            add(bit1, r+1, -x,N)
        else:
            res =sum(bit0,r) + sum(bit1, r)*r
            res-=sum(bit0,l-1) + sum(bit1, l-1) * (l-1)
            print(res)

if __name__ == '__main__':
    A=[0,5,3,7,9,6,4,1,2]
    N=8
    Q=2
    T=["C","Q"]
    L=[1,1]
    R=[5,8]
    X=[1,0]
    solve(A,N,Q,T,L,R,X)
